src/gptq_cuda.rs

```diff
@@ -0,0 +1,180 @@
+    let q = QTransformerWeightsCuda {
+        rms_eps: sel.rms_eps,
+        token_embedding_table: sel.token_embedding_table,
+        rms_att_weight: sel.rms_att_weight.into_iter().collect(),
+        wq: sel.wq.iter().map(|x| x.convert().expect("works")).collect(),
+        wk: sel.wk.iter().map(|x| x.convert().expect("works")).collect(),
+        wv: sel.wv.iter().map(|x| x.convert().expect("works")).collect(),
+        wo: sel.wo.iter().map(|x| x.convert().expect("works")).collect(),
+        rms_ffn_weight: sel.rms_ffn_weight.into_iter().collect(),
+        w1: sel.w1.iter().map(|x| x.convert().expect("works")).collect(),
+        w2: sel.w2.iter().map(|x| x.convert().expect("works")).collect(),
+        w3: sel.w3.iter().map(|x| x.convert().expect("works")).collect(),
+        rms_final_weight: sel.rms_final_weight,
+        _freq_cis_real: sel._freq_cis_real,
+        _freq_cis_imag: sel._freq_cis_imag,
+        wcls: sel.wcls,
+    };
+    Ok(q)
+}
+
+impl<
+        const IN: usize,
+        const OUT: usize,
+        const GROUPS: usize,
+        const ING: usize,
+        const OUTG: usize,
+    > QLinear<IN, OUT, GROUPS, ING, OUTG>
+{
+    pub fn convert(self: &Self) -> Result<QLinearCuda<IN, OUT, GROUPS, ING, OUTG>, Box<dyn Error>> {
+        let t: [F16; OUT] = [F16(f16::from_f32(0.0)); OUT];
+        let out = QLinearCuda {
+            qweight: self.qweight.as_dbuf()?,
+            scales: self.scales.as_dbuf()?,
+            qzeros: self.qzeros.as_dbuf()?,
+            g_index: self.g_index.clone(),
+            xout_temp: t.as_dbuf()?,
+            module: Module::from_ptx(PTX, &[])?,
+        };
+        Ok(out)
+    }
+}
+#[derive(Debug)]
+pub struct QLinearCuda<
+    const IN: usize,
+    const OUT: usize,
+    const GROUPS: usize,
+    const ING: usize,
+    const OUTG: usize,
+> {
+    // The quantized weights
+    qweight: DeviceBuffer<[i32; ING]>,
+    // The zero term per group.
+    qzeros: DeviceBuffer<[i32; GROUPS]>,
+    // The scale term per group
+    scales: DeviceBuffer<[f32; GROUPS]>,
+    // Remapping for ACT_ORDER=True
+    pub g_index: [i32; IN],
+    xout_temp: DeviceBuffer<F16>,
+    module: Module,
+}
+
+impl<
+        const IN: usize,
+        const OUT: usize,
+        const GROUPS: usize,
+        const ING: usize,
+        const OUTG: usize,
+    > QLinearCuda<IN, OUT, GROUPS, ING, OUTG>
+{
+    pub fn matvec<const B: usize>(self: &Self, xout: &mut [[f32; OUT]; B], x: &[[f32; IN]; B]) {
+        assert_eq!(ING, IN / 32 * BITS);
+        assert_eq!(OUTG, OUT / 32 * BITS);
+        assert_eq!(B, 1);
+
+        let mut x_temp: [[F16; IN]; B] = [[F16(f16::from_f32(0.0)); IN]; B];
+        // Remap the input for ACT_ORDER=True
+        for i in 0..B {
+            for j in 0..IN {
+                x_temp[i][j] = F16(f16::from_f32(x[i][self.g_index[j] as usize]));
+            }
+        }
+
+        let res = || {
+            let xtemp_gpu = x_temp[0].as_dbuf()?;
```

Comment:
Each time you call matvec, you do CPU->GPU copy. This operations is very slow, even if you copy only small amount of data, you still incur transfer overheads.

Reply:
Yeah, my plan was to first try fusing qkv attention and MLP layer. If I copy DIM values 2 times per layer, GPU should be fast enough. (But we'll see).